### DISCUSSION TITLE
Makefile-Probleme - missing ')'...

### DISCUSSION THREAD

**Discussion by Bodenseematze** [Freetz-NG/freetz-ng#584] [Allgemein] [answered]
Hallo,

ich wollte mal wieder meine Images aktualisieren und neue bauen...
...dazu habe ich wie immer "git pull" und dann "make oldconfig" gemacht.

Da kommt leider eine Fehlermeldung:
`freetz@freetz64:~/workdir/7490$ make oldconfig`
`Freetz-NG 20037M-37d64c4f7 master 2022-07-08`
`Makefile:226: *** unterminated call to function 'shell': missing ')'.  Stop.`
`make: *** [Makefile:46: envira] Terminated`

Die Meldung kommt auch, wenn ich make ohne Parameter aufrufe...

Ich muss leider gestehen, dass ich keinen Fehler in der Zeile 226 entdecken kann...

**Comment by fda77** [comment]
Sehr komisch. hast du evtl noch weitere eigene lokale  Änderungen? -> `git status`
Dieser "AVX" Block ist recht neu: https://github.com/Freetz-NG/freetz-ng/blob/master/Makefile#L223=
Hast du eine CPU mit AVX(1)? Das gibt es seit "Sandy Bridge" , also Gen2 iCore
Du könntest einzelne Zeilen auskommentieren um die betreffende zu finden.

**Reply by Bodenseematze** [reply]
Nein, ich habe keine lokalen Änderungen - "git status" zeigt alles "clean" an.
Ich habe noch einen anderen ausgecheckten Clone, der ist auf "Freetz-NG 19784-06beb5e14 master 2022-06-16" - da ist noch alles OK.

Das ganze wird verwendet in einer virtuellen Maschine (Ubuntu 20.04) unter VirtualBox.
Spielt da AVX eine Rolle (was ist das überhaupt)?

Das hier kommt bei `cat /proc/cpuinfo`:
`processor       : 0`
`vendor_id       : GenuineIntel`
`cpu family      : 6`
`model           : 140`
`model name      : 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz`
`stepping        : 1`
`microcode       : 0xffffffff`
`cpu MHz         : 2726.604`
`cache size      : 12288 KB`
`physical id     : 0`
`siblings        : 1`
`core id         : 0`
`cpu cores       : 1`
`apicid          : 0`
`initial apicid  : 0`
`fpu             : yes`
`fpu_exception   : yes`
`cpuid level     : 22`
`wp              : yes`
`flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx rdtscp lm `
`constant_tsc rep_good nopl xtopology nonstop_tsc cpuid pni ssse3 cx16 pcid sse4_1 sse4_2 hypervisor lahf_lm invpcid_single`
`ibrs_enhanced fsgsbase invpcid md_clear flush_l1d arch_capabilities`
`bugs            : spectre_v1 spectre_v2 spec_store_bypass swapgs`
`bogomips        : 5453.20`
`clflush size    : 64`
`cache_alignment : 64`
`address sizes   : 39 bits physical, 48 bits virtual`
`power management:`

EDIT: ich habe jetzt mal die Version des Makefile aus dem älteren Build mit dem aktuellen verglichen - die einzigen Zeilen, die hinzugefügt wurden, sind die für die AVX-Prüfung.
Wenn ich die entferne, funktioniert es wieder ohne Probleme.

Ich kann allerdings immer noch keinen Grund für die Fehlermeldung sehen - die Zeilen werden in ähnlicher Form überall im Makefile verwendet - warum machen gerade diese jetzt Probleme?

**Comment by fda77** [comment]
Du hast also eine recht neue CPU die AVX, AVX2 und sogar AVX-512 kann, aber deine VM-Software reicht diese und andere Features nicht durch. Vielleicht solltest du das beheben, das hab aber nichts mit dem Problem hier zu tun. Das steht dann in der "flags" Zeile.

Ich habe die Zeile geändert
```
-ifneq ($(shell grep -q ' avx ' /proc/cpuinfo && echo avx),avx)
+ifneq ($(shell grep -q ' avx ' /proc/cpuinfo && echo avxx),xxxxxx)
```
und bekomme dann wie erwartet
```
make
Freetz-NG 20060MO-7e354ba2e master 2022-07-11
You have no CPU with AVX support, precompiled (download) toolchains automatically disabled.
You have no CPU with AVX support, precompiled (download) host-tools automatically disabled.
Makefile:263: *** Please re-run.  Schluss.
```
> Ich kann allerdings immer noch keinen Grund für die Fehlermeldung sehen - die Zeilen werden in ähnlicher Form überall im Makefile verwendet - warum machen gerade diese jetzt Probleme?

Keine Ahnung! Da bei mir der Fehler nicht auftritt müsstest du es herausfinden. Kommentier doch mal die einzelnen `DLCHG` Zeilen aus oder die `$(info` um genau zu wissen welche das auslöst

Welches Linux benutzt du eigentlich?

**Comment by Bodenseematze** [comment]
Wie oben geschrieben - es handelt sich um Ubuntu 20.04 in einer VM, bereitgestellt über VirtualBox.

Ich habe keine Ahnung, wie bzw ob es überhaupt möglich ist, VirtualBox dazu zu überreden, die Flags an die VM weiter zu reichen.

Wenn ich es richtig verstanden habe, dann sorgen die neuen Abfragen dafür, dass ohne (erkanntem) AVX die vorkompilierte Toolchain deaktiviert wird. Korrekt?
Die Verwendung der vorkompilierten Toolchain hat bei mir eigentlich schon immer funktioniert - mit oder ohne (erkanntem) AVX :-)

**Comment by fda77** [comment]
> Ich habe keine Ahnung, wie bzw ob es überhaupt möglich ist, VirtualBox dazu zu überreden, die Flags an die VM weiter zu reichen.

Hab da einen Geheimtipp: google.com

>  Korrekt?

Genau

> hat bei mir eigentlich schon immer funktioniert

Gähn, ich kann es nicht mehr lesen. So ein alberner "Argument".
Ich hab keine Lust meine Zeit mit nem uralten Ubuntu 14  zu verschwenden um die Sachen zu erzeugen, der Aufwand wird immer größer bei 0 Nutzen.
Wer eine mehr als 10 Jahre alte CPU ohne AVX hat, bekommt halt die vorkompilierten Sachen deaktiviert. Selbst compilieren geht ja, und mit so einer alten CPU ist man eh warten gewöhnt.

Hast du herausfinden können wo das Problem genau herkommt?

**Reply by Bodenseematze** [reply]
Tja, Google hilft nur bedingt - anscheinend kann man Virtualbox zwar prinzipiell dazu überreden, die AVX-Features an die VMs weiter zu reichen...
...aber anscheinend wohl nicht, wenn auf dem Rechner noch Hyper-V aktiv ist - was bei mir der Fall ist und was ich auch nicht ändern kann/mag, weil ich Hyper-V für andere Dinge brauche...

Wieso uraltes Ubuntu 14? Ich habe das "offizielle" Freetz-Image verwendet, das ist 64-Bit Ubuntu 20.04 (LTS) - das sehe ich nicht als "uralt" an!
Und mit dem Satz, dass es bei mir immer  schon funktioniert hat, meinte ich, dass die Toolchain in meiner VM funktioniert hat, obwol dort der AVX-Support offensichtlich nicht von VirtualBox weitergereicht wurde - und wenn ich die AVX-Prüfung im Script auskommentiere, funktioniert die Toolbox immer noch.

Und nein, ich habe noch nicht herausgefunden wo das Problem genau herkommt.
Selbst wenn ich "nur" ein "echo Hallo" in die Zeilen über (!) der AVX-Prüfung hinzufüge, kommt der Fehler - dann in der "echo"-Zeile...
...und nach wie vor: sobald ich den Block mit der AVX-Prüfung auskommentiere, geht alles...
...ich kann auch die ifneq/ifeq-Zeilen (samt zugehörigen endif) drin lassen, solange ich die "DLGCHG:="-Zeilen auskommentiert lasse.
Aber sobald ich da eine Zeile nicht mehr auskommentiert habe, schlägt das mit der besagten Fehlermeldung fehl...
EDIT: es scheint, als ob Du das Problem gefunden hast - die letzte Version des Makefile mit "\\#" in den sed-Zeilen funktioniert wieder bei mir (halt mit Erkennung, dass ich kein AVX habe - aber dann baue ich halt die Toolchain selber...)! Super!

**Reply by fda77** [reply]
Wenn dein System kein AVX kann, musst du halt die Toolchains selbst ohne AVX bauen.
Manches nutzt keine der AVX Instruktionen und funktioniert deshalb, ich wünsche viel Glück. Wenn du daran herumbastelst öffne aber bitte keine Issues hier!
Ubuntu 14: Lies einfach oben nochmal

**Comment by fda77** [comment, marked as answer]
fixed: https://github.com/Freetz-NG/freetz-ng/commit/45e3d4b1e99d67fceec7d3b850b598a68d182380 thx @meiser79 !
Komisch dass das bei Fedora keinen Fehler verursacht
